src/remote.rs

```diff
@@ -1,117 +1,142 @@
+                    mime::TEXT => Ok(mtype),
                     _ => Err(AxoassetError::RemoteAssetNonImageMimeType {
-                        asset: self.to_string(),
+                        origin_path: origin_path.to_string(),
                     }),
                 }
             }
             None => Err(AxoassetError::RemoteAssetMissingContentTypeHeader {
-                asset: self.to_string(),
+                origin_path: origin_path.to_string(),
             }),
         }
     }
 
-    fn extension(&self) -> Result<String> {
-        let mimetype = self.mimetype()?;
+    fn extension(mimetype: mime::Mime, origin_path: &str) -> Result<String> {
+        match mimetype.type_() {
+            mime::IMAGE => RemoteAsset::image_extension(mimetype, origin_path),
+            mime::TEXT => RemoteAsset::text_extension(mimetype, origin_path),
+            _ => Err(AxoassetError::RemoteAssetMimeTypeNotSupported {
+                origin_path: origin_path.to_string(),
+                mimetype: mimetype.to_string(),
+            }),
+        }
+    }
+
+    fn text_extension(mimetype: mime::Mime, origin_path: &str) -> Result<String> {
+        if let Some(extension) = mimetype.suffix() {
+            Ok(extension.to_string())
+        } else {
+            match mimetype.subtype() {
+                mime::PLAIN => Ok("txt".to_string()),
+                mime::CSS => Ok("css".to_string()),
+                _ => Err(AxoassetError::RemoteAssetMimeTypeNotSupported {
+                    origin_path: origin_path.to_string(),
+                    mimetype: mimetype.to_string(),
+                }),
+            }
+        }
+    }
+
+    fn image_extension(mimetype: mime::Mime, origin_path: &str) -> Result<String> {
         if let Some(img_format) = image::ImageFormat::from_mime_type(&mimetype) {
             let extensions = img_format.extensions_str();
             if !extensions.is_empty() {
                 Ok(extensions[0].to_string())
             } else {
                 Err(
                     AxoassetError::RemoteAssetIndeterminateImageFormatExtension {
-                        asset: self.to_string(),
+                        origin_path: origin_path.to_string(),
                     },
                 )
             }
         } else {
             Err(AxoassetError::RemoteAssetMimeTypeNotSupported {
-                asset: self.to_string(),
+                origin_path: origin_path.to_string(),
                 mimetype: mimetype.to_string(),
             })
         }
     }
 
-    fn filename(&self) -> Result<String> {
-        Ok(format!("logo.{}", self.extension()?))
+    fn filename(origin_path: &str, headers: &reqwest::header::HeaderMap) -> Result<String> {
+        let mut filestem = url::Url::parse(origin_path)?
+            .path()
+            .to_string()
+            .replace('/', "_");
+        filestem.remove(0);
+        if filestem.contains('.') {
+            Ok(filestem)
```

Comment:
genuinely confused by this logic... it seems like this does the following series of transforms?

1. https://mywebsite.com/some/path/image.png
2. /some/path/image.png
3. _some_path_image.png
4. some_path_image.png

Is that... really what you want? You want to keep the full path as the image name (maybe useful for avoiding conflicts)? But not any part of the hostname (adding conflicts back in...)?

Reply:
I know you said that was jank you'd clean up later but at very least this merits a comment for the current behaviour and what problems it's solving/failing-to-solve.

Reply:
will add, i didn't think about the hostname removing conflicts- i was trying to get the shortest possible filename that was as likely as possible to remove conflicts so was thinking i would codify "the path" which in my head didn't include the hostname- i'd be down to re-add it or alternatively consider a different filename structure. hashing the file and appending to the filename/final url segment could also work?

Reply:
filed #6 and #9 and added a comment to the function in this PR